sdk/python/feast/offline_store.py

```diff
@@ -0,0 +1,140 @@
+        start_date: datetime,
+        end_date: datetime,
+    ) -> pyarrow.Table:
+
+        partition_by_entity_string = ", ".join(entity_names)
+        if partition_by_entity_string != "":
+            partition_by_entity_string = "PARTITION BY " + partition_by_entity_string
+        timestamps = [event_timestamp_column]
+        if created_timestamp_column is not None:
+            timestamps.append(created_timestamp_column)
+        timestamp_desc_string = " DESC, ".join(timestamps) + " DESC"
+        field_string = ", ".join(entity_names + feature_names + timestamps)
+
+        query = f"""
+        SELECT {field_string}
+        FROM (
+            SELECT {field_string},
+            ROW_NUMBER() OVER({partition_by_entity_string} ORDER BY {timestamp_desc_string}) AS _feast_row
+            FROM `{table_ref}`
+            WHERE {event_timestamp_column} BETWEEN TIMESTAMP('{start_date}') AND TIMESTAMP('{end_date}')
+        )
+        WHERE _feast_row = 1
+        """
+        return BigQueryOfflineStore._pull_query(query)
+
+    @staticmethod
+    def _pull_query(query: str) -> pyarrow.Table:
+        from google.cloud import bigquery
+
+        client = bigquery.Client()
+        query_job = client.query(query)
+        return query_job.to_arrow()
+
+
+def run_reverse_field_mapping(
+    feature_view: FeatureView,
+) -> Tuple[List[str], List[str], str, Optional[str]]:
+    # if we have mapped fields, use the original field names in the call to the offline store
+    event_timestamp_column = feature_view.inputs.event_timestamp_column
+    entity_names = [entity.name for entity in feature_view.entities]
+    feature_names = [feature.name for feature in feature_view.features]
+    created_timestamp_column = feature_view.inputs.created_timestamp_column
+    if feature_view.inputs.field_mapping is not None:
+        reverse_field_mapping = {
+            v: k for k, v in feature_view.inputs.field_mapping.items()
+        }
+        event_timestamp_column = (
+            reverse_field_mapping[event_timestamp_column]
+            if event_timestamp_column in reverse_field_mapping.keys()
+            else event_timestamp_column
+        )
+        created_timestamp_column = (
+            reverse_field_mapping[created_timestamp_column]
+            if created_timestamp_column is not None
+            and created_timestamp_column in reverse_field_mapping.keys()
+            else created_timestamp_column
+        )
+        entity_names = [
+            reverse_field_mapping[col] if col in reverse_field_mapping.keys() else col
+            for col in entity_names
+        ]
+        feature_names = [
+            reverse_field_mapping[col] if col in reverse_field_mapping.keys() else col
+            for col in feature_names
+        ]
+    return (
+        entity_names,
+        feature_names,
+        event_timestamp_column,
+        created_timestamp_column,
+    )
+
+
+def run_forward_field_mapping(
```

Comment:
What about `forward_map_columns`? (map as verb)

Reply:
I'd like to keep it like this to match the other function